crypto/src/merkle.rs

```diff
@@ -9,262 +9,256 @@ use iroha_schema::prelude::*;
+
+    fn get(&self, idx: usize) -> Option<&T>;
+
+    /// Get the reference of the `idx`-th leaf node.
+    fn get_leaf(&self, idx: usize) -> Option<&T> {
+        let offset = 2_usize.pow(self.height()) - 1;
+        offset.checked_add(idx).and_then(|i| self.get(i))
     }
-    fn schema(map: &mut MetaMap) {
-        map.entry(Self::type_name()).or_insert_with(|| {
-            // BFS ordered list of leaf nodes
-            Metadata::Vec(VecMeta {
-                ty: HashOf::<T>::type_name(),
-                sorted: true,
-            })
-        });
-        if !map.contains_key(&HashOf::<T>::type_name()) {
-            HashOf::<T>::schema(map);
+
+    fn height(&self) -> u32 {
+        (usize::BITS - self.len().leading_zeros()).saturating_sub(1)
+    }
+
+    fn max_nodes_at_height(&self) -> usize {
+        2_usize.pow(self.height() + 1) - 1
+    }
+
+    fn parent(&self, idx: usize) -> Option<usize> {
+        if 0 == idx {
+            return None;
         }
+        let idx = (idx - 1).div_euclid(2);
+        (idx < self.len()).then(|| idx)
     }
-}
 
-/// Represents a subtree rooted by the current node.
-#[derive(Debug)]
-struct Subtree<T> {
-    /// Left child node.
-    left: Box<Node<T>>,
-    /// Right child node.
-    right: Box<Node<T>>,
-    /// Hash of this node.
-    hash: Option<HashOf<Node<T>>>,
-}
+    fn l_child(&self, idx: usize) -> Option<usize> {
+        let idx = 2 * idx + 1;
+        (idx < self.len()).then(|| idx)
+    }
 
-/// Represents a leaf node.
-#[derive(Debug)]
-struct Leaf<T> {
-    /// Hash of this node.
-    hash: HashOf<T>,
-}
+    fn r_child(&self, idx: usize) -> Option<usize> {
+        let idx = 2 * idx + 2;
+        (idx < self.len()).then(|| idx)
+    }
 
-/// Binary tree node: [`Subtree`], [`Leaf`] (with data or links to data), or `Empty`.
-#[derive(Debug)]
-#[allow(clippy::module_name_repetitions)]
-enum Node<T> {
-    /// [`Subtree`] node.
-    Subtree(Subtree<T>),
-    /// [`Leaf`] node.
-    Leaf(Leaf<T>),
-    /// Empty node.
-    Empty,
+    fn get_l_child(&self, idx: usize) -> Option<&T> {
+        self.l_child(idx).and_then(|i| self.get(i))
+    }
+
+    fn get_r_child(&self, idx: usize) -> Option<&T> {
+        self.r_child(idx).and_then(|i| self.get(i))
+    }
 }
 
-#[derive(Debug)]
-/// BFS iterator over [`MerkleTree`].
-struct BreadthFirstNodeIterator<'itm, T> {
-    queue: VecDeque<&'itm Node<T>>,
+impl<T> CompleteBTree<Option<HashOf<T>>> for MerkleTree<T> {
+    fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    fn get(&self, idx: usize) -> Option<&Option<HashOf<T>>> {
+        self.0.get(idx)
+    }
 }
 
 #[cfg(feature = "std")]
 impl<T> FromIterator<HashOf<T>> for MerkleTree<T> {
     fn from_iter<I: IntoIterator<Item = HashOf<T>>>(iter: I) -> Self {
-        let mut nodes = iter
-            .into_iter()
-            .map(|hash| Node::Leaf(Leaf { hash }))
-            .collect::<VecDeque<_>>();
-        nodes.make_contiguous().sort_unstable_by_key(Node::hash);
```

Comment:
The sorting of leaves is not implemented in either `MerkleTree::from_iterator` or `MerkleTree::add`.
AFAICT there is no need to keep the root hash unique to a SET of leaves

Reply:
> I only see `Vec` around the merkle trees, but would there still be a problem?

`Vec` does guarantee ordering, and there is no `Ord` or `Eq` which would benefit from this (maybe there should be?).

> In that case I have to revert the `MerkleTree::add` optimization.

Our tree is being created out of iterators in the code. I don't think we're ever inserting just one element

Hypothetical case: there is a `HashSet<T>` out of which you create two `MerkleTree<T>`. You sign them with the same key but you find that the signatures don't match

I'm not pushing for sorting per se. It's a tradeoff we should discuss

Reply:
One more thing. When creating from an iterator complexity should be O(nlog(n)). Therefore sorting will not affect complexity

Reply:
Yes, I've corrected myself

Reply:
If `add` weren't necessary we could have the nodes sorted without paying any price. Do we need merkletree in kura?

Reply:
https://github.com/hyperledger/iroha/issues/2388

Reply:
I had a change of heart. I think it's better to preserve insertion order. You can reslove this as far as I am concerned

Reply:
I guess we can't reach consensus after all.

Reply:
Added a comment https://github.com/hyperledger/iroha/issues/2388#issuecomment-1166194214
Anyway keeping the insertion order improves (not completely purify) `Handler<GetBlockHash>::handle` since this was totally broken before the PR -- it was getting an ordering in hashes for the block height request